eventhubs-streamanalytics-eventhubs/tools/eh-asa-perfmon/EventReceiver.cs

```diff
@@ -0,0 +1,119 @@
+﻿using Azure.Messaging.EventHubs.Consumer;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace StreamingAtScale
+{
+    public class EventReceiver
+    {
+        private readonly object _lock = new object();
+
+        public async Task Receive(string connectionString, CancellationToken cancellationToken)
+        {
+            // In here, our consumer will read from the latest position instead of the earliest.  As a result, it won't see events that
+            // have previously been published. 
+            //
+            // Each partition of an Event Hub represents potentially infinite stream of events.  When a consumer is reading, there is no definitive
+            // point where it can assess that all events have been read and no more will be available.  As a result, when the consumer reaches the end of
+            // the available events for a partition, it will continue to wait for new events to arrive so that it can surface them to be processed. During this
+            // time, the iterator will block.
+            //
+            // In order to prevent the consumer from waiting forever for events, and blocking other code, there is a method available for developers to
+            // control this behavior. It's signaling the cancellation token passed when reading will cause the consumer to stop waiting and end iteration
+            // immediately.  This is desirable when you have decided that you are done reading and do not wish to continue.  
+
+            await using (var consumerClient = new EventHubConsumerClient(EventHubConsumerClient.DefaultConsumerGroupName, connectionString))
+            {
+                Console.WriteLine("The application will now start to listen for incoming message.");
+
+                // Each time the consumer looks to read events, we'll ask that it waits only a short time before emitting
+                // an empty event, so that our code has the chance to run without indefinite blocking.
+
+                ReadEventOptions readOptions = new ReadEventOptions
+                {
+                    MaximumWaitTime = TimeSpan.FromMilliseconds(500)
+                };
+
+                using (var csvOutput = File.CreateText("./result.csv"))
+                {
+                    csvOutput.WriteLine("EventCount,BatchCount,BatchFrom,BatchTo,MinLatency,MaxLatency,AvgLatency");
+                    try
+                    {
+                        // The client is safe and intend to be long-lived; in order to begin reading at the
+                        // latest event, we'll need to specify that reading should not start at earliest.
+
+                        var events = consumerClient.ReadEventsAsync(startReadingAtEarliestEvent: false, readOptions, cancellationToken);
```

Comment:
We need to be a bit cautious when describing `startReadingAtEarliestEvent: false`.   It does not start reading at the latest event, but rather will only see events that are published after you've started reading.   It's a subtle difference, but one that isn't intuitive.

To illustrate, when line 53 executes, the reader will attach to all partitions but will not see anything that is already there.  If no new events are published, the consumer will read nothing.  Only when a new event is published while the `foreach` loop is active, will anything be read.  I don't see anything in this sample that is doing active publishing, so if there's not an associated ASA instance pushing events, this code will never read.

Unless you have a clear need to not read from the beginning of each partition, I'd recommend using the default behavior and allowing the sample to start reading at the beginning.

Reply:
Yes, i got some misunderstanding. Now it's much more clear.  Thanks